### PR TITLE
feat: add peer block filter option

### PR DESCRIPTION
**Pull request by laurentsenta**
This feature lets a user configure a filter to allow / deny bitswap requests according to the request's peer ID and the content id. For example, a user may use this option to implement a dynamic peer-based authorization.

Contributes to https://github.com/ipfs/go-ipfs/issues/8763,
Relates to https://github.com/ipfs/go-bitswap/pull/549

This PR adds the plumbing required to pass the peer-block filter option from ipfs down to go-bitswap.

- [ ] confirm with the team the way we pass the option is the right approach,
- [ ] confirm with the team where we want to document this,
   - we could advertise this as an experimental feature,
   - provide an example / demo of the feature in the docs/examples folder (like in [this demo](https://github.com/laurentsenta/go-ipfs/tree/feat/peer-block-filter-with-demo/docs/examples/go-ipfs-as-a-library))

### CODE REVIEW

File: core/node/builder.go
Comment (laurentsenta) [inline core/node/builder.go:46]: I'm not sure which we should prefer:
1. Pass the `PeerBlockRequestFilter` function, shown here,
2. Pass a generic `bitswap.Option` function.

(2) is more "obviously future-proof", but that'd mean we'd have two different ways to configure most values in the bitswap instance  (json from cfg and code from buildcfg).

Access control is probably a cross-cutting concern that will span multiple modules and protocols, not only bitswap. So by decoupling the access control config from the bitswap option, (1) might make it easier to build upon this feature.

File: core/node/groups.go
Comment (Jorropo) [inline core/node/groups.go:287]: I belive you inspired your pbrf variable from the line above, this code (`shouldBitswapProvide`) is a clarification variable, it translate `!cfg.Experimental.StrategicProviding` into a usefull thing we understand.
In your case you are turning a perfectly good `PeerBlockRequestFilter` into a meaningless without prior knowledge `pbrf` one.
```suggestion
	return fx.Options(
		fx.Provide(OnlineExchange(cfg, shouldBitswapProvide, bcfg.PeerBlockRequestFilter)),
```

**Review by Jorropo — CHANGES_REQUESTED** [review]
Your code doesn't build:
```
core/node/bitswap.go:27:60: undefined: bitswap.PeerBlockRequestFilter
core/node/bitswap.go:45:24: undefined: bitswap.WithPeerBlockRequestFilter
core/node/builder.go:46:25: undefined: bitswap.PeerBlockRequestFilter
```
I guess you are waiting on a new version of bitswap with your changes merged.
I don't know if maintainers have feelings about this, but I like to create a temporary replace rule that I let maintainers remove (with the "maintainers are allowed to push to this branch" or alike option).
You can do this by running:
```
go mod edit -replace <target repo>=<your fork>@<your commit>
```
For example:
```
go mod edit -replace github.com/alanshaw/go-carbites=github.com/Jorropo/go-carbites@751dcbf09c8a387030238e4bd3169cc2516046f2
```
You can also add a `// see <link to the blocking pull request>` comment on the replace line in the `go.mod`.

That allows the CI tests to run and shows that your code is RFM (blocking PR aside).

**Comment by laurentsenta** [comment]
Closing this after a sync chat with Adin, this might come back as a plugin later.

Thanks for the initial review @Jorropo, I won't update the PR but I'll keep your feedback in mind for future PRs.

**Pull request closed without merging**
